Incorrect tokenizing

Found a bug where the string `bredcp` is tokenized as
```rust
Token { ty: Ident, source: "br", range: 0..2 }
Token { ty: Ident, source: "bredcp", range: 0..6 }
```
Instead of 
```rust
Token {
    ty: TokenType::Ident,
    source: "bredcp",
    range: (0, 6)
}
```
The ident token is [here](https://github.com/Kixiron/crunch-lang/blob/dev/crates/crunch-parser/src/token.rs#L16-L17), and it's extra confusing because I don't have any tokens that are remotely close to `br`, plus the fact that it's being somehow *re*tokenized (As indicated by the ranges of the tokens makes this even stranger.
I also made a [test case](https://github.com/Kixiron/crunch-lang/blob/dev/crates/crunch-parser/src/token.rs#L443-L456) that exemplifies this error. I do have a small iterator adapter wrapped around logos, but this shouldn't have any effect as all it does is filter out unused tokens

I switched versions and have been greeted with 48 of these:
```rust
error[E0308]: mismatched types
 --> crates\crunch-parser\src\token.rs:6:10
  |
6 | #[derive(Logos, Debug, PartialEq, Eq, Clone, Copy, Hash)]
  |          ^^^^^ expected fn item, found a different fn item
  |
  = note: expected fn item `for<'r> fn(&'r mut logos::lexer::Lexer<token::TokenType, _>) {<token::TokenType as logos::Logos>::lexicon::_handle_0_newline::<'_, _>}`
             found fn item `for<'r> fn(&'r mut logos::lexer::Lexer<token::TokenType, _>) {<token::TokenType as logos::Logos>::lexicon::_error::<'_, _>}`
```

--- a/logos-derive/Cargo.toml
+++ b/logos-derive/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "logos-derive"
-version = "0.10.0-rc3"
+version = "0.10.0-rc5"
 authors = ["Maciej Hirsz <hello@maciej.codes>"]
 license = "MIT/Apache-2.0"
 description = "Create ridiculously fast Lexers"

--- a/logos-derive/Cargo.toml
+++ b/logos-derive/Cargo.toml
@@ -18,7 +18,6 @@ proc-macro = true
 [dependencies]
 beef = "0.4.1"
 fnv = "1.0.6"
-rustc-hash = "1.0.1"
 syn = { version = "1.0.17", features = ["extra-traits", "full"] }
 quote = "1.0.3"
 proc-macro2 = "1.0.9"

--- a/logos-derive/src/graph/mod.rs
+++ b/logos-derive/src/graph/mod.rs
@@ -129,6 +129,10 @@ impl<Leaf> Graph<Leaf> {
             (None, None) => {
                 panic!("Merging two reserved nodes!");
             },
+            // Merging a leaf with an empty slot would produce
+            // an empty self-referencing fork.
+            (Some(Node::Leaf(_)), None) => return a,
+            (None, Some(Node::Leaf(_))) => return b,
             (Some(Node::Leaf(left)), Some(Node::Leaf(right))) => {
                 return match Disambiguate::cmp(left, right) {
                     Ordering::Less => b,
@@ -172,8 +176,30 @@ impl<Leaf> Graph<Leaf> {
         }
 
         let mut fork = self.fork_off(a);
-
         fork.merge(self.fork_off(b), self);
+
+        let mut stack = vec![id.get()];
+
+        // Flatten the fork
+        while let Some(miss) = fork.miss {
+            if stack.contains(&miss) {
+                break;
+            }
+            stack.push(miss);
+
+            match self.get(miss) {
+                Some(Node::Fork(other)) => {
+                    match other.miss {
+                        Some(id) if self.get(id).is_none() => break,
+                        _ => (),
+                    }
+                    fork.miss = None;
+                    fork.merge(other.clone(), self);
+
+                },
+                _ => break,
+            }
+        }
 
         self.insert(id, fork)
     }

--- a/logos-derive/src/graph/regex.rs
+++ b/logos-derive/src/graph/regex.rs
@@ -6,7 +6,7 @@ use regex_syntax::hir::{Class, ClassUnicode, Hir, HirKind, Literal, RepetitionKi
 use regex_syntax::ParserBuilder;
 use utf8_ranges::Utf8Sequences;
 
-use crate::graph::{Graph, Disambiguate, NodeId, Range, Rope, Fork};
+use crate::graph::{Graph, Disambiguate, Node, NodeId, Range, Rope, Fork};
 use crate::error::{Error, Result};
 
 /// Middle Intermediate Representation of the regex, built from
@@ -115,13 +115,20 @@ impl<Leaf: Disambiguate + Debug> Graph<Leaf> {
         match mir {
             Mir::Empty => (0, then),
             Mir::Loop(mir) => {
+                let miss = match miss {
+                    Some(id) => self.merge(id, then),
+                    None => then,
+                };
                 let this = self.reserve();
-                let (_, id) = self.parse_mir(*mir, this.get(), miss);
-                let id = self.merge(id, then);
+                let (_, id) = self.parse_mir(*mir, this.get(), Some(miss));
 
                 // Move the node to the reserved id
-                let fork = self.fork_off(id);
-                let id = self.insert(this, fork);
+                let node = match &self[id] {
+                    Node::Fork(fork) => Node::Fork(fork.clone()),
+                    Node::Rope(rope) => Node::Rope(rope.clone()),
+                    Node::Leaf(_) => unreachable!(),
+                };
+                let id = self.insert(this, node);
 
                 (0, id)
             }

--- a/logos/Cargo.toml
+++ b/logos/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "logos"
-version = "0.10.0-rc4"
+version = "0.10.0-rc5"
 authors = ["Maciej Hirsz <hello@maciej.codes>"]
 license = "MIT/Apache-2.0"
 description = "Create ridiculously fast Lexers"
@@ -12,7 +12,7 @@ readme = "../README.md"
 edition = "2018"
 
 [dependencies]
-logos-derive = { version = "0.10.0-rc3", optional = true }
+logos-derive = { version = "0.10.0-rc5", optional = true }
 
 [features]
 default = ["export_derive", "std"]

--- a/tests/tests/crunch.rs
+++ b/tests/tests/crunch.rs
@@ -1,0 +1,31 @@
+use logos_derive::Logos;
+
+#[derive(Logos, Debug, Clone, Copy, PartialEq)]
+enum Token {
+    #[end]
+    End,
+    #[error]
+    Error,
+    #[token = "else"]
+    Else,
+    #[token = "exposed"]
+    Exposed,
+    #[regex = "[^ \t\n\r\"\'!@#$%\\^&*()-+=,.<>/?;:\\[\\]{}\\\\|`~]+"]
+    Ident,
+
+}
+
+mod crunch {
+    use super::*;
+    use tests::assert_lex;
+
+    #[test]
+    fn crunch() {
+        assert_lex(
+            "exposed_function",
+            &[
+                (Token::Ident, "exposed_function", 0..16),
+            ],
+        );
+    }
+}

